api/grpcserver.go

```diff
@@ -285,11 +285,27 @@ func (svr *GRPCServer) ReadState(ctx context.Context, in *iotexapi.ReadStateRequ
 
 // EstimateGasForAction estimates gas for action
 func (svr *GRPCServer) EstimateGasForAction(ctx context.Context, in *iotexapi.EstimateGasForActionRequest) (*iotexapi.EstimateGasForActionResponse, error) {
-	estimateGas, err := svr.coreService.EstimateGasForAction(in.Action)
+	selp, err := (&action.Deserializer{}).ActionToSealedEnvelope(in.Action)
 	if err != nil {
-		return nil, err
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	callerAddr := selp.SrcPubkey().Address()
+	if callerAddr == nil {
+		return nil, status.Error(codes.Internal, "failed to get address")
+	}
+	sc, ok := selp.Action().(*action.Execution)
+	if !ok {
+		gas, err := selp.IntrinsicGas()
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return &iotexapi.EstimateGasForActionResponse{Gas: gas}, nil
+	}
+	_, receipt, err := svr.coreService.SimulateExecution(ctx, callerAddr, sc)
```

Comment:
1. in gasstation.go, it is `bc.Context(ctx)`
2. as discussed, all these changes should go to `func (core *coreService) EstimateGasForAction()`, this file `grpcserver.go` should *not* change at all

Reply:
ok. revert